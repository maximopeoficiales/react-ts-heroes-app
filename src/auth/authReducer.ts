export enum Action {
    LOGIN = "login",
    LOGOUT = "logout",
    DEFAULT = "",
}
export type User =
    {
        name?: string;
        logged?: boolean;
    }

export type ReducerUser = { type: Action, payload?: User };

export const authReducer = (state: User, action: ReducerUser): User => {
    switch (action.type) {
        case Action.LOGIN:
            return { ...action.payload, logged: true }
        case Action.LOGOUT:
            return { logged: false };
        default:
            return state;
    }

}