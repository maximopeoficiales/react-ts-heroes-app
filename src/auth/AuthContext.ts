import { createContext } from "react";
import { ReducerUser, User } from "./authReducer";

export interface UserProvider {
    user?: User,
    dispatch: React.Dispatch<ReducerUser>
}
export const AuthContext = createContext<UserProvider>({ dispatch: () => { } });