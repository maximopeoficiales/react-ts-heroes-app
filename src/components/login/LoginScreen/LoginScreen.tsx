import { useContext } from "react";
import { useForm } from "../../../hooks/useForm";
import { AuthContext } from "../../../auth/AuthContext";
import { Action, User } from "../../../auth/authReducer";
import Container from "../../utils/Container/Container";
import { History } from "history";

// de esta manera extiendo de RouteComponentProps
interface MyProps {
  history: History;
}

const defaultProps = {};
const LoginScreen = (props: MyProps) => {
  props = { ...defaultProps, ...props };
  const { history } = props;

  const { formValues, handlerChange } = useForm<{ username: string }>({
    username: "",
  });
  const { dispatch } = useContext(AuthContext);
  const { username } = formValues;
  const handlerSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    
    if (username.trim() === "") {
      return;
    }
    // cuando haga login redireccionara a la ruta que quizo acceder
    const lastPath = localStorage.getItem("lastPath") || "/";
    const userLogin: User = { name: username, logged: true };

    dispatch({ type: Action.LOGIN, payload: userLogin });
    history.replace(lastPath);
  };

  return (
    <div data-testid="LoginScreen" className="text-white">
      <Container>
        <h2 className="my-4 text-3xl font-bold text-center">Login</h2>
        <form method="post" onSubmit={handlerSubmit}>
          <div className="flex flex-col my-2 ">
            <label htmlFor="username" className="text-base font-bold">
              Insert to User
            </label>
            <input
              value={username}
              onChange={handlerChange}
              type="text"
              className="p-2 my-1 border-none rounded-lg outline-none bg-heroBlack text-heroWhite"
              name="username"
              autoComplete="off"
              id="username"
              placeholder="Example: user123"
            />
          </div>

          <button
            type="submit"
            // onClick={}
            className="w-full px-3 py-2 font-bold duration-75 ease-out delay-75 transform scale-100 bg-green-600 rounded-full active:scale-75 active:bg-green-800 "
          >
            Login
          </button>
        </form>
      </Container>
    </div>
  );
};

export default LoginScreen;
