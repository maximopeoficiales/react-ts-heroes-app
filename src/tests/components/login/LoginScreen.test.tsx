import { mount } from "enzyme";
import React from "react";
import { AuthContext, UserProvider } from "../../../auth/AuthContext";
import { Action, ReducerUser } from "../../../auth/authReducer";
import LoginScreen from "../../../components/login/LoginScreen/LoginScreen";

describe("<LoginScreen />", () => {
  const history = {
    push: jest.fn(),
    replace: jest.fn(),
    location: {} as any,
    listen: jest.fn(),
    createHref: jest.fn(),
  };
  const contextValue: UserProvider = {
    dispatch: jest.fn(),
    user: {
      logged: false,
    },
  };
  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <LoginScreen history={history as any} />
    </AuthContext.Provider>
  );
  test("it should mount", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de realizar el dispatch y la navegacion", () => {
    let name = "maximo";
    wrapper
      .find("input")
      .simulate("change", { target: { value: name, name: "username" } });
    const formSubmit: any = wrapper.find("form").prop("onSubmit");

    // simula evento de submit
    formSubmit({ preventDefault() {} });

    const action: ReducerUser = {
      type: Action.LOGIN,
      payload: { logged: true, name },
    };
    expect(contextValue.dispatch).toHaveBeenCalledWith(action);
    expect(history.replace).toHaveBeenCalledWith("/");
    // si existiera el lastPath
    localStorage.setItem("lastPath", "/dc");
    formSubmit({ preventDefault() {} });
    expect(history.replace).toHaveBeenCalledWith("/dc");
  });
});
