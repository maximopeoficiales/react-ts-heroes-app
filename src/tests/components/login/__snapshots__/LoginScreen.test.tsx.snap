// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> it should mount 1`] = `
<div
  className="text-white"
  data-testid="LoginScreen"
>
  <div
    className="w-11/12 mx-auto my-8 lg:w-9/12"
  >
    <div
      className="p-5 rounded-xl lg:p-14 bg-heroBlackSecondary"
    >
      <h2
        className="my-4 text-3xl font-bold text-center"
      >
        Login
      </h2>
      <form
        method="post"
        onSubmit={[Function]}
      >
        <div
          className="flex flex-col my-2 "
        >
          <label
            className="text-base font-bold"
            htmlFor="username"
          >
            Insert to User
          </label>
          <input
            autoComplete="off"
            className="p-2 my-1 border-none rounded-lg outline-none bg-heroBlack text-heroWhite"
            id="username"
            name="username"
            onChange={[Function]}
            placeholder="Example: user123"
            type="text"
            value=""
          />
        </div>
        <button
          className="w-full px-3 py-2 font-bold duration-75 ease-out delay-75 transform scale-100 bg-green-600 rounded-full active:scale-75 active:bg-green-800 "
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</div>
`;
